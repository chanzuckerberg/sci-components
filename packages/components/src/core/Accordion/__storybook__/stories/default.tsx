import { Args } from "@storybook/react";
import RawAccordion, {
  AccordionDetails,
  AccordionHeader,
} from "src/core/Accordion";
import { EXTRA_SHORT_LOREM_IPSUM } from "src/common/storybook/loremIpsum";

export const Accordion = (props: Args): JSX.Element => {
  const { id, subtitle, useDivider, togglePosition } = props;
  return (
    <RawAccordion
      id={id}
      useDivider={useDivider}
      togglePosition={togglePosition}
      {...props}
    >
      <AccordionHeader id={`${id}-header`} subtitle={subtitle}>
        Accordion Header
      </AccordionHeader>
      <AccordionDetails>{EXTRA_SHORT_LOREM_IPSUM}</AccordionDetails>
    </RawAccordion>
  );
};
