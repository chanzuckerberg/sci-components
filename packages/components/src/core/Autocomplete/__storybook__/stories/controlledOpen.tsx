import { Args } from "@storybook/types";
import { DefaultAutocompleteOption } from "src/core/Autocomplete/components/AutocompleteBase";
import { AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS } from "src/common/storybook/AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS";
import { useState } from "react";
import RawAutocomplete from "src/core/Autocomplete";
import { AUTOCOMPLETE_WRAPPER_STYLES } from "../constants";

export const ControlledOpenDemo = <T extends DefaultAutocompleteOption>(
  props: Args
): JSX.Element => {
  const {
    label,
    multiple,
    options = AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS,
  } = props;

  const [open, setOpen] = useState(true);

  return (
    <div style={AUTOCOMPLETE_WRAPPER_STYLES}>
      <RawAutocomplete
        onClickAway={() => {
          setOpen(false);
        }}
        onClick={() => {
          if (open) {
            setOpen(false);
          } else {
            setOpen(true);
          }
        }}
        open={open}
        id="autocomplete-demo"
        disableCloseOnSelect={multiple}
        label={label}
        multiple={multiple}
        options={options}
        getOptionDisabled={(option: T) => {
          return option.name === "Type: feature request";
        }}
        {...props}
      />
    </div>
  );
};
