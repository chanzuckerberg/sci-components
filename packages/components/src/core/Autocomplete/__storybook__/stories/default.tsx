import { Args } from "@storybook/react";
import { DefaultAutocompleteOption } from "src/core/Autocomplete/components/AutocompleteBase";
import { AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS } from "src/common/storybook/AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS";
import { SyntheticEvent, useEffect, useState } from "react";
import RawAutocomplete, {
  AutocompleteMultiColumnValue,
  SDSAutocompleteValue,
} from "src/core/Autocomplete";
import TagFilter from "src/core/TagFilter";
import { AutocompleteValue } from "@mui/base";
import {
  AUTOCOMPLETE_DIV_MARGIN,
  AUTOCOMPLETE_WRAPPER_STYLES,
} from "../constants";

export const Autocomplete = <
  T extends DefaultAutocompleteOption,
  Multiple extends boolean | undefined,
>(
  props: Args
): JSX.Element => {
  const {
    label,
    multiple,
    options = AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS,
    search,
    value: propValue,
    keepSearchOnSelect,
  } = props;

  type DisableClearable = false;
  type FreeSolo = false;

  const isControlled = propValue !== undefined;
  const [value, setValue] = useState<
    SDSAutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  >(
    (multiple ? [] : null) as SDSAutocompleteValue<
      T,
      Multiple,
      DisableClearable,
      FreeSolo
    >
  );

  const [selection, setSelection] = useState<string[]>([]);

  useEffect(() => {
    if (isControlled) {
      setValue(propValue);
    }
  }, [isControlled, propValue]);

  useEffect(() => {
    setSelection([]);
    setValue(
      (multiple ? [] : null) as SDSAutocompleteValue<
        T,
        Multiple,
        DisableClearable,
        FreeSolo
      >
    );
  }, [multiple, options]);

  return (
    <>
      <div style={AUTOCOMPLETE_WRAPPER_STYLES}>
        <RawAutocomplete
          id="autocomplete-demo"
          disableCloseOnSelect={multiple}
          label={label}
          multiple={multiple}
          onChange={handleChange}
          keepSearchOnSelect={keepSearchOnSelect}
          options={options}
          search={search}
          value={value}
          getOptionDisabled={(option: T) => {
            return option.name === "Type: feature request";
          }}
          {...props}
        />
      </div>
      <div style={{ margin: AUTOCOMPLETE_DIV_MARGIN }}>
        {selection.length
          ? selection.map((item) => {
              return (
                <TagFilter
                  key={item}
                  label={item}
                  onDelete={() => handleTagDelete(item)}
                  onClick={() => handleTagDelete(item)}
                />
              );
            })
          : null}
      </div>
    </>
  );

  // eslint-disable-next-line sonarjs/cognitive-complexity
  function handleChange(
    _: SyntheticEvent<Element, Event>,
    newValue: SDSAutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  ) {
    setValue(newValue);
    const newSelection: string[] = [];

    if (
      (newValue as T)?.name ||
      (Array.isArray(newValue) && (newValue as T[])[0].name)
    ) {
      if (multiple) {
        setSelection(
          Array.isArray(newValue) ? newValue?.map((item) => item.name) : []
        );
      } else {
        if (newValue && !Array.isArray(newValue) && newValue.name) {
          setSelection([(newValue as T).name]);
        }
      }
    } else {
      if (multiple) {
        if (newValue) {
          Object.values(newValue).forEach((items) => {
            (items as T[])?.map(({ name }: { name: string }) => {
              newSelection.push(name);
            });
          });
          setSelection(newSelection);
        }
      } else {
        if (newValue) {
          Object.values(newValue).forEach((item) => {
            newSelection.push((item as T).name);
          });
        }
        setSelection(newSelection);
      }
    }
  }

  // eslint-disable-next-line sonarjs/cognitive-complexity
  function handleTagDeleteMultiColumn(tag: string) {
    if (multiple) {
      if (value) {
        Object.entries(value).forEach(([key, values]) => {
          const index = (values as T[])?.findIndex(
            (item: T) => item.name === tag
          );
          if (index > -1) {
            const newValue = [...(values as T[])];
            newValue.splice(index, 1);
            setValue((prev) => {
              return {
                ...prev,
                [key]: newValue as AutocompleteValue<
                  T,
                  Multiple,
                  DisableClearable,
                  FreeSolo
                >,
              };
            });

            const newSelection = [...selection];
            const deleteIndex = newSelection.indexOf(tag);
            newSelection.splice(deleteIndex, 1);
            setSelection(newSelection);
          }
        });
      }
    } else {
      if (value) {
        Object.entries(value).forEach(([key, val]) => {
          if ((val as T).name === tag) {
            const finalValue = { ...value } as AutocompleteMultiColumnValue<
              T,
              Multiple,
              DisableClearable,
              FreeSolo
            >;
            if (finalValue) delete finalValue[key];
            setValue(finalValue);

            const newSelection = [...selection];
            const deleteIndex = newSelection.indexOf(tag);
            newSelection.splice(deleteIndex, 1);
            setSelection(newSelection);
          }
        });
      }
    }
  }

  function handleTagDelete(tag: string) {
    if (
      (value && (value as T).name) ||
      (Array.isArray(value) && (value as T[])[0].name)
    ) {
      if (multiple && Array.isArray(value)) {
        const index = value?.findIndex((item) => item.name === tag);
        const newValue = [...value];
        newValue.splice(index, 1);
        setValue(
          newValue as SDSAutocompleteValue<
            T,
            Multiple,
            DisableClearable,
            FreeSolo
          >
        );

        const newSelection = [...selection];
        const deleteIndex = newSelection.indexOf(tag);
        newSelection.splice(deleteIndex, 1);
        setSelection(newSelection);
      } else {
        setValue(
          null as SDSAutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
        );
        setSelection([]);
      }
    } else {
      handleTagDeleteMultiColumn(tag);
    }
  }
};
