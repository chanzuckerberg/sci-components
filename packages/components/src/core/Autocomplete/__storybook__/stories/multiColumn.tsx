import { Args } from "@storybook/react";
import { DefaultAutocompleteOption } from "src/core/Autocomplete/components/AutocompleteBase";
import { AUTOCOMPLETE_MULTI_COLUMN_OPTIONS } from "src/common/storybook/AUTOCOMPLETE_MULTI_COLUMN_OPTIONS";
import { useEffect, useState } from "react";
import { AutocompleteMultiColumnValue } from "src/core/Autocomplete";
import {
  AutocompleteChangeDetails,
  AutocompleteChangeReason,
  AutocompleteValue,
} from "@mui/base";
import {
  AUTOCOMPLETE_DIV_MARGIN,
  AUTOCOMPLETE_WRAPPER_STYLES,
} from "../constants";
import TagFilter from "src/core/TagFilter";
import RawAutocompleteMultiColumn from "src/core/Autocomplete/components/AutocompleteMultiColumn";

export const AutocompleteMultiColumnDemo = <
  T extends DefaultAutocompleteOption,
  Multiple extends boolean | undefined,
>(
  props: Args
): JSX.Element => {
  const {
    multiple,
    options = AUTOCOMPLETE_MULTI_COLUMN_OPTIONS,
    search,
    label,
  } = props;

  type DisableClearable = false;
  type FreeSolo = false;

  const [selection, setSelection] = useState<string[]>([]);

  const [theValue, setTheValue] =
    useState<
      AutocompleteMultiColumnValue<T, Multiple, DisableClearable, FreeSolo>
    >();

  useEffect(() => {
    setSelection([]);
  }, [multiple]);

  const handleChange = (
    _event: React.SyntheticEvent,
    newValue: AutocompleteMultiColumnValue<
      T,
      Multiple,
      DisableClearable,
      FreeSolo
    >,
    _reason: AutocompleteChangeReason,
    _details?: AutocompleteChangeDetails<T>
  ) => {
    setTheValue(newValue);
    const newSelection: string[] = [];

    if (multiple) {
      if (newValue) {
        Object.values(newValue).forEach((items) => {
          (items as T[])?.map(({ name }: { name: string }) => {
            newSelection.push(name);
          });
        });
        setSelection(newSelection);
      }
    } else {
      if (newValue) {
        Object.values(newValue).forEach((item) => {
          newSelection.push((item as T).name);
        });
      }
      setSelection(newSelection);
    }
  };

  return (
    <>
      <div style={AUTOCOMPLETE_WRAPPER_STYLES}>
        <RawAutocompleteMultiColumn<T, Multiple, DisableClearable, FreeSolo>
          label={label}
          disableCloseOnSelect={false}
          multiple={multiple}
          onChange={handleChange}
          options={options}
          search={search}
          value={theValue}
          {...props}
        />
      </div>
      <div style={{ margin: AUTOCOMPLETE_DIV_MARGIN }}>
        {selection.length
          ? selection.map((item) => {
              return (
                <TagFilter
                  key={item}
                  label={item}
                  onDelete={() => handleTagDelete(item)}
                  onClick={() => handleTagDelete(item)}
                />
              );
            })
          : null}
      </div>
    </>
  );

  function handleTagDelete(tag: string) {
    if (multiple) {
      if (theValue) {
        Object.entries(theValue).forEach(([key, values]) => {
          const index = (values as T[])?.findIndex(
            (item: T) => item.name === tag
          );
          if (index > -1) {
            const newValue = [...(values as T[])];
            newValue.splice(index, 1);
            setTheValue((prev) => {
              return {
                ...prev,
                [key]: newValue as AutocompleteValue<
                  T,
                  Multiple,
                  DisableClearable,
                  FreeSolo
                >,
              };
            });

            const newSelection = [...selection];
            const deleteIndex = newSelection.indexOf(tag);
            newSelection.splice(deleteIndex, 1);
            setSelection(newSelection);
          }
        });
      }
    } else {
      if (theValue) {
        Object.entries(theValue).forEach(([key, value]) => {
          if ((value as T).name === tag) {
            const finalValue = { ...theValue };
            delete finalValue[key];
            setTheValue(finalValue);

            const newSelection = [...selection];
            const deleteIndex = newSelection.indexOf(tag);
            newSelection.splice(deleteIndex, 1);
            setSelection(newSelection);
          }
        });
      }
    }
  }
};
