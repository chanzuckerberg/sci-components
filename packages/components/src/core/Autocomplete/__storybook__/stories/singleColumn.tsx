import { AutocompleteChangeReason, AutocompleteValue } from "@mui/base";
import AutocompleteBase, {
  DefaultAutocompleteOption,
} from "src/core/Autocomplete/components/AutocompleteBase";
import { Args } from "@storybook/react";
import { AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS } from "src/common/storybook/AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS";
import { useEffect, useState } from "react";
import {
  AUTOCOMPLETE_DIV_MARGIN,
  AUTOCOMPLETE_WRAPPER_STYLES,
} from "../constants";
import TagFilter from "src/core/TagFilter";

export const AutocompleteSingleColumnDemo = <
  T extends DefaultAutocompleteOption,
  Multiple extends boolean | undefined,
>(
  props: Args
): JSX.Element => {
  type DisableClearable = false;
  type FreeSolo = false;

  const {
    label,
    multiple,
    options = AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS,
    search,
    value: propValue,
    keepSearchOnSelect,
  } = props;

  const isControlled = propValue !== undefined;
  const [value, setValue] = useState(getInitialValue());
  const [pendingValue, setPendingValue] = useState(getInitialValue());

  const [selection, setSelection] = useState<string[]>([]);

  useEffect(() => {
    if (isControlled) {
      setValue(propValue);
    }
  }, [isControlled, propValue]);

  useEffect(() => {
    setSelection([]);
  }, [multiple]);

  return (
    <>
      <div style={AUTOCOMPLETE_WRAPPER_STYLES}>
        <AutocompleteBase
          id="autocomplete-base-demo"
          disableCloseOnSelect={multiple}
          label={label}
          multiple={multiple}
          keepSearchOnSelect={keepSearchOnSelect}
          options={options}
          search={search}
          onChange={handleChange}
          value={multiple ? pendingValue : value}
          getOptionDisabled={(option: DefaultAutocompleteOption) => {
            return option.name === "Type: feature request";
          }}
          {...props}
        />
      </div>
      <div style={{ margin: AUTOCOMPLETE_DIV_MARGIN }}>
        {selection.length
          ? selection.map((item) => {
              return (
                <TagFilter
                  key={item}
                  label={item}
                  onDelete={() => handleTagDelete(item)}
                  onClick={() => handleTagDelete(item)}
                />
              );
            })
          : null}
      </div>
    </>
  );

  function handleChange(
    _event: React.SyntheticEvent,
    newValue: AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>,
    _reason: AutocompleteChangeReason
  ) {
    if (multiple) {
      const newSelection = Array.isArray(newValue)
        ? newValue?.map((item) => item.name)
        : [];
      setSelection(newSelection);
      return setPendingValue(newValue);
    } else {
      if (newValue && !Array.isArray(newValue) && newValue.name) {
        setValue(newValue);
        setSelection([newValue.name]);
      }
    }
  }

  function handleTagDelete(tag: string) {
    if (multiple && Array.isArray(pendingValue)) {
      const index = pendingValue?.findIndex((item) => item.name === tag);
      const newValue = [...pendingValue];
      newValue.splice(index, 1);
      setPendingValue(
        newValue as AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
      );

      const newSelection = [...selection];
      const deleteIndex = newSelection.indexOf(tag);
      newSelection.splice(deleteIndex, 1);
      setSelection(newSelection);
    } else {
      setValue(
        null as unknown as AutocompleteValue<
          T,
          Multiple,
          DisableClearable,
          FreeSolo
        >
      );
      setSelection([]);
    }
  }

  function getInitialValue(): AutocompleteValue<
    T,
    Multiple,
    DisableClearable,
    FreeSolo
  > {
    if (isControlled) {
      return propValue;
    }

    return multiple
      ? ([] as unknown as AutocompleteValue<
          T,
          Multiple,
          DisableClearable,
          FreeSolo
        >)
      : (null as unknown as AutocompleteValue<
          T,
          Multiple,
          DisableClearable,
          FreeSolo
        >);
  }
};
