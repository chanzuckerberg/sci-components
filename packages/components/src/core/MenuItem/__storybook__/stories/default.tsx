import { Args } from "@storybook/react";
import RawMenuItem from "src/core/MenuItem";
import { DemoWrapper } from "../style";

export const MenuItem = (props: Args): JSX.Element => {
  const { name } = props;
  return (
    <DemoWrapper>
      <RawMenuItem data-testid="MenuItem" {...props}>
        {name}
      </RawMenuItem>
    </DemoWrapper>
  );
};
