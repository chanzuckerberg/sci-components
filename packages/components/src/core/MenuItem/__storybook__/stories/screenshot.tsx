import { Args } from "@storybook/react";
import {
  MENU_ITEM_COLUMN_OPTIONS,
  MENU_ITEM_DISABLED_OPTIONS,
  MENU_ITEM_MULTI_SELECT_OPTIONS,
  MENU_ITEM_PSEUDO_STATES,
  MENU_ITEM_SCREENSHOT_ICON_OPTIONS,
  MENU_ITEM_SDS_STYLE_OPTIONS,
  MENU_ITEM_SELECTED_OPTIONS,
} from "../constants";
import RawMenuItem, { MenuItemProps } from "src/core/MenuItem";

export const ScreenshotTestDemo = (props: Args): JSX.Element => {
  const DISPLAY_CONTENTS: React.CSSProperties = {
    display: "contents",
  };
  const MID_LABEL: React.CSSProperties = {
    borderStyle: "solid none none none",
    gridColumn: "1 / 6",
    justifySelf: "stretch",
    paddingTop: 10,
  };

  // loop through all MULTI_SELECT_OPTIONS
  return (
    <>
      {MENU_ITEM_MULTI_SELECT_OPTIONS.map((isMultiSelect) => {
        return (
          <MenuItemMultiSelect
            isMultiSelect={isMultiSelect}
            key={String(isMultiSelect)}
          />
        );
      })}
    </>
  );

  // loop through all COLUMN_OPTIONS + create headers for MULTI_SELECT_OPTIONS
  function MenuItemMultiSelect({
    isMultiSelect,
  }: {
    isMultiSelect: (typeof MENU_ITEM_MULTI_SELECT_OPTIONS)[number];
  }) {
    const LEVEL_STYLE: React.CSSProperties = {
      columnGap: "20px",
      display: "inline-grid",
      fontFamily: "sans-serif",
      marginRight: "50px",
    };
    const LABEL_STYLE: React.CSSProperties = {
      fontSize: "2em",
      gridColumn: "1 / 6",
      marginBottom: 10,
    };

    return (
      <div style={LEVEL_STYLE}>
        <p style={LABEL_STYLE}>
          Multi-select: <b>{isMultiSelect ? "yes" : "no"}</b>
        </p>
        {MENU_ITEM_COLUMN_OPTIONS.map((column) => {
          return (
            <MenuItemColumn
              isMultiSelect={isMultiSelect}
              column={column}
              key={String(column)}
            />
          );
        })}
      </div>
    );
  }

  // loop through all ICON_OPTIONS + create headers for COLUMN_OPTIONS
  function MenuItemColumn({
    isMultiSelect,
    column,
  }: {
    isMultiSelect: (typeof MENU_ITEM_MULTI_SELECT_OPTIONS)[number];
    column: (typeof MENU_ITEM_COLUMN_OPTIONS)[number];
  }) {
    const LABEL_STYLE: React.CSSProperties = {
      ...MID_LABEL,
      alignSelf: "end",
      borderWidth: "3px",
      fontSize: "1.17em",
      margin: "0 0 15px 0",
    };
    return (
      <div style={DISPLAY_CONTENTS}>
        <p style={LABEL_STYLE}>
          Column: <b>{column ? "yes" : "no"}</b>
        </p>
        {MENU_ITEM_SCREENSHOT_ICON_OPTIONS.map((sdsIcon) => {
          return (
            <MenuItemIcon
              isMultiSelect={isMultiSelect}
              column={column}
              sdsIcon={sdsIcon}
              key={String(sdsIcon)}
            />
          );
        })}
      </div>
    );
  }

  // loop through all SELECTED_OPTIONS + create headers for ICON_OPTIONS
  function MenuItemIcon({
    isMultiSelect,
    column,
    sdsIcon,
  }: {
    isMultiSelect: (typeof MENU_ITEM_MULTI_SELECT_OPTIONS)[number];
    column: (typeof MENU_ITEM_COLUMN_OPTIONS)[number];
    sdsIcon: (typeof MENU_ITEM_SCREENSHOT_ICON_OPTIONS)[number];
  }) {
    const LABEL_STYLE: React.CSSProperties = {
      ...MID_LABEL,
      alignSelf: "end",
      borderWidth: "2px",
      fontSize: "1em",
      margin: "0 0 15px 0",
    };
    return (
      <div style={DISPLAY_CONTENTS}>
        <p style={LABEL_STYLE}>
          Icon: <b>{sdsIcon ? "yes" : "no"}</b>
        </p>
        {MENU_ITEM_SELECTED_OPTIONS.map((selected) => {
          return (
            <MenuItemSelected
              isMultiSelect={isMultiSelect}
              column={column}
              sdsIcon={sdsIcon}
              selected={selected}
              key={String(selected)}
            />
          );
        })}
      </div>
    );
  }

  // loop through all DISABLED_OPTIONS + create headers for SELECTED_OPTIONS
  function MenuItemSelected({
    isMultiSelect,
    column,
    sdsIcon,
    selected,
  }: {
    isMultiSelect: (typeof MENU_ITEM_MULTI_SELECT_OPTIONS)[number];
    column: (typeof MENU_ITEM_COLUMN_OPTIONS)[number];
    sdsIcon: (typeof MENU_ITEM_SCREENSHOT_ICON_OPTIONS)[number];
    selected: (typeof MENU_ITEM_SELECTED_OPTIONS)[number];
  }) {
    const LABEL_STYLE: React.CSSProperties = {
      ...MID_LABEL,
      alignSelf: "end",
      borderWidth: "1px",
      fontSize: "0.83em",
      margin: "0 0 10px 0",
    };
    return (
      <div style={DISPLAY_CONTENTS}>
        <p style={LABEL_STYLE}>
          Selected: <b>{selected ? "true" : "false"}</b>
        </p>
        {MENU_ITEM_SDS_STYLE_OPTIONS.map((sdsStyle) => {
          return (
            <MenuItemStyles
              isMultiSelect={isMultiSelect}
              column={column}
              sdsIcon={sdsIcon}
              selected={selected}
              sdsStyle={sdsStyle}
              key={String(sdsStyle)}
            />
          );
        })}
      </div>
    );
  }

  // loop through all SDS_STYLE_OPTIONS + create headers for DETERMINISTIC_OPTIONS
  function MenuItemStyles({
    isMultiSelect,
    column,
    sdsIcon,
    selected,
    sdsStyle,
  }: {
    isMultiSelect: (typeof MENU_ITEM_MULTI_SELECT_OPTIONS)[number];
    column: (typeof MENU_ITEM_COLUMN_OPTIONS)[number];
    sdsIcon: (typeof MENU_ITEM_SCREENSHOT_ICON_OPTIONS)[number];
    selected: (typeof MENU_ITEM_SELECTED_OPTIONS)[number];
    sdsStyle: (typeof MENU_ITEM_SDS_STYLE_OPTIONS)[number];
  }) {
    const LABEL_STYLE: React.CSSProperties = {
      ...MID_LABEL,
      alignSelf: "end",
      borderWidth: "1px",
      fontSize: "0.83em",
      margin: "0 0 10px 0",
    };
    return (
      <div style={DISPLAY_CONTENTS}>
        <p style={LABEL_STYLE}>
          SdsStyle: <b>{sdsStyle}</b>
        </p>
        {MENU_ITEM_DISABLED_OPTIONS.map((disabled) => {
          return (
            <MenuItemDisabled
              isMultiSelect={isMultiSelect}
              column={column}
              sdsIcon={sdsIcon}
              selected={selected}
              sdsStyle={sdsStyle as MenuItemProps<"Gear">["sdsStyle"]}
              disabled={disabled}
              key={String(disabled)}
            />
          );
        })}
      </div>
    );
  }

  // loop through all PSEUDO_STATES + create headers for DISABLED_OPTIONS, PSEUDO_STATES
  function MenuItemDisabled({
    isMultiSelect,
    column,
    sdsIcon,
    selected,
    sdsStyle,
    disabled,
  }: {
    isMultiSelect: (typeof MENU_ITEM_MULTI_SELECT_OPTIONS)[number];
    column: (typeof MENU_ITEM_COLUMN_OPTIONS)[number];
    sdsIcon: (typeof MENU_ITEM_SCREENSHOT_ICON_OPTIONS)[number];
    selected: (typeof MENU_ITEM_SELECTED_OPTIONS)[number];
    sdsStyle: MenuItemProps<"Gear">["sdsStyle"];
    disabled: (typeof MENU_ITEM_DISABLED_OPTIONS)[number];
  }) {
    const LEVEL_STYLE: React.CSSProperties = {
      marginBottom: 20,
    };
    const LABEL_STYLE: React.CSSProperties = {
      fontSize: "0.67em",
      margin: "10px 0 20px 0",
    };
    return (
      <div style={DISPLAY_CONTENTS}>
        {MENU_ITEM_PSEUDO_STATES.map((state) => {
          return (
            // to pass a11y, this container div must have role of "menu" since it contains components with roles of "menuitem"
            <div role="menu" style={LEVEL_STYLE} key={state}>
              {/* removes irrelevant disabled iterations: when combined with all pseudo-states except default, `disabled=false` is impossible */}
              {(disabled === false ||
                (disabled === true && state === "default")) && (
                <>
                  <p style={LABEL_STYLE}>
                    {disabled ? "Disabled: " : "State: "}
                    <b>{disabled ? "true" : state}</b>
                  </p>
                  <RawMenuItem
                    {...props}
                    data-testid="menu-item"
                    isMultiSelect={isMultiSelect}
                    column={column}
                    icon={sdsIcon}
                    selected={selected}
                    sdsStyle={sdsStyle}
                    disabled={disabled}
                    className={`pseudo-${state}`}
                    key={state}
                  >
                    Name
                  </RawMenuItem>
                </>
              )}
            </div>
          );
        })}
      </div>
    );
  }
};
