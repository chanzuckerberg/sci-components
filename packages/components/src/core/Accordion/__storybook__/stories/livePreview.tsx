import { Args } from "@storybook/react";
import { Accordion } from "./default";

const livePreviewStyles = {
  display: "grid",
  gridColumnGap: "24px",
  gridRowGap: "50px",
  gridTemplateColumns: "repeat(4, 200px)",
};

export function LivePreviewDemo(props: Args): JSX.Element {
  return (
    <div style={livePreviewStyles as React.CSSProperties}>
      <Accordion
        id="accordion-1"
        togglePosition="right"
        useDivider
        {...props}
      />
      <Accordion
        id="accordion-2"
        togglePosition="right"
        subtitle="Optional Subtitle"
        useDivider
        {...props}
      />
      <Accordion id="accordion-3" togglePosition="left" useDivider {...props} />
      <Accordion
        id="accordion-4"
        togglePosition="left"
        subtitle="Optional Subtitle"
        useDivider
        {...props}
      />
    </div>
  );
}
