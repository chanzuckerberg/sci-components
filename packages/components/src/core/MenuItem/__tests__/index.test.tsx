import { generateSnapshots } from "@chanzuckerberg/story-utils";
import { composeStories } from "@storybook/react";
import { render, screen } from "@testing-library/react";
import * as stories from "../__storybook__/index.stories";

// Returns a component that already contain all decorators from story level, meta level and global level.
const { Test } = composeStories(stories);

describe("<MenuItem />", () => {
  generateSnapshots(stories);

  it("renders menu item component", () => {
    render(<Test />);
    const elements = screen.getAllByTestId("MenuItem");
    expect(elements).toBeTruthy();
  });
});
