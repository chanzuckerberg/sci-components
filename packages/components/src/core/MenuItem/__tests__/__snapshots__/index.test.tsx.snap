// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuItem /> Default story renders snapshot 1`] = `
<div
  class="css-1m75yyh"
>
  <li
    class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-pelx1i-MuiButtonBase-root-MuiMenuItem-root"
    data-testid="MenuItem"
    name="Text here"
    role="menuitem"
    tabindex="-1"
  >
    <span
      class="css-bv5fko"
    >
      <span
        class="primary-text css-1k1xg99"
      >
        Text here
      </span>
      <span
        class="css-1o15svb"
      >
        123,456
      </span>
    </span>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </li>
</div>
`;
